engine/crates/engine/value/src/lib.rs

```diff
@@ -643,3 +648,34 @@ fn write_object<K: Display, V: Display>(
     }
     f.write_char('}')
 }
+
+/// Iterator over the variables that are used inside a Value
+pub struct VariableIterator<'a> {
+    values: Vec<&'a Value>,
```

Comment:
Why not `&'a [&'a Value]` together with `position: usize`. Your iterator returns references anyhow, so using a slice here would allow the whole struct to be `Copy`.

Reply:
How would I recurse through nested objects & lists if I had a slice?

Reply:
Maybe add `nested: Option<VariableIterator<'a>>` here and then first do `nested.and_then(|iter| iter.next())` and if it returns something, return that immediately.

Reply:
I can't use a slice either, since I'm sometimes pulling from an `IndexMap<String, Value>`.

I'm sure I could do it with a

```
struct VariableIterator<'a> {
   values: Box<dyn Iterator<Item=Value>>,
   nested: Option<Box<VariableIterator<'a>>
}
```

or something, but I kind of like the simplicity of the current and I'm not too fussed about allocations